File: app/src/main/java/com/niedzwiecki/przemyslguide/ui/login/password/PasswordViewModel.java
package com.niedzwiecki.przemyslguide.ui.login.password;

import android.databinding.ObservableField;
import android.databinding.ObservableInt;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;

import com.niedzwiecki.przemyslguide.data.DataManager;
import com.niedzwiecki.przemyslguide.data.StringManager;
import com.niedzwiecki.przemyslguide.ui.base.BaseViewModel;
import com.niedzwiecki.przemyslguide.ui.base.Navigator;
import com.niedzwiecki.przemyslguide.util.Utils;

import javax.inject.Inject;

import rx.Subscription;

public class PasswordViewModel extends BaseViewModel {

    private static final int START_EMAIL_ACTIVITY = 0;
    private static final int START_PASSWORD_ACTIVITY = 1;

    DataManager dataManager;
    public ObservableField<String> passwordFromEditText;
    public ObservableField<String> validationPasswordErrorText;
    public ObservableInt errorVisibility;
    public ObservableInt nextButtonVisibility;
    private String passwordEditText;
    private String emailAddress;
    private Subscription mSubscrition;
    private ObservableField<String> error;

    @Inject
    public PasswordViewModel(DataManager dataManager) {
        this.dataManager = dataManager;
        this.validationPasswordErrorText = new ObservableField<>();
        this.error = new ObservableField<>();
        this.passwordFromEditText = new ObservableField<>();
        errorVisibility = new ObservableInt(View.INVISIBLE);
        nextButtonVisibility = new ObservableInt(View.GONE);
    }

    public TextWatcher getPasswordTextWatcher() {
        return new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                passwordEditText = charSequence.toString();
                passwordFromEditText.set(passwordEditText);
                validPassword(passwordEditText);
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        };
    }

    public void validPassword(CharSequence password) {
        if (Utils.isEmpty(password)) {
            showValidationErrorMessage(dataManager.
                    getResourcesManager().getString(StringManager.ERROR_EMPTY_PASSWORD_VALIDATION));
        } else if (!Utils.isPasswordLengthMinimumFourChars(password)) {
            showValidationErrorMessage(dataManager.
                    getResourcesManager().getString(StringManager.ERROR_NOT_ENOUGH_PASSWORD_CHARS));
        } else {
            nextButtonVisibility.set(View.VISIBLE);
            errorVisibility.set(View.INVISIBLE);
        }
    }

    private void showValidationErrorMessage(String errorMessage) {
        validationPasswordErrorText.set(errorMessage);
        errorVisibility.set(View.VISIBLE);
        nextButtonVisibility.set(View.GONE);
    }

    public void startMainActivity(View view) {
        dataManager.storeAuthenticationHeader(emailAddress);
        getNavigator().moveForward(Navigator.Options.START_MAIN_ACTIVITY, emailAddress);
        getNavigator().finish();
    }

    public void sendEmailToVerification(String email) {
        emailAddress = email;
    }

    public void checkIfUserIsVerified() {
        if (isNavigatorAttached()) {
            dataManager.storeAuthenticationHeader(emailAddress);
            getNavigator().moveForward(Navigator.Options.START_MAIN_ACTIVITY, emailAddress);
        }
    }

}

File: app/src/main/java/com/niedzwiecki/przemyslguide/ui/login/password/PasswordViewModel.kt
package com.niedzwiecki.przemyslguide.ui.login.password

import android.databinding.ObservableField
import android.databinding.ObservableInt
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import com.niedzwiecki.przemyslguide.R
import com.niedzwiecki.przemyslguide.data.DataManager
import com.niedzwiecki.przemyslguide.ui.base.BaseViewModel
import com.niedzwiecki.przemyslguide.ui.base.Navigator
import com.niedzwiecki.przemyslguide.util.Utils

class PasswordViewModel(internal var dataManager: DataManager) : BaseViewModel<PasswordActivity>() {
    var passwordFromEditText: ObservableField<String>
    var validationPasswordErrorText: ObservableField<String>
    var errorVisibility: ObservableInt
    var nextButtonVisibility: ObservableInt
    private var passwordEditText: String? = null
    private var emailAddress: String? = null
    private val error: ObservableField<String>

    val passwordTextWatcher: TextWatcher
        get() = object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                passwordEditText = charSequence.toString()
                passwordFromEditText.set(passwordEditText)
                validPassword(passwordEditText)
            }

            override fun afterTextChanged(editable: Editable) {

            }
        }

    init {
        this.validationPasswordErrorText = ObservableField()
        this.error = ObservableField()
        this.passwordFromEditText = ObservableField()
        errorVisibility = ObservableInt(View.INVISIBLE)
        nextButtonVisibility = ObservableInt(View.GONE)
    }

    fun validPassword(password: CharSequence?) {
        if (Utils.isEmpty(password)) {
            showValidationErrorMessage(
                    dataManager.getResourcesManager().getString(
                            R.string.emptyPasswordValidation
                    )
            )
        } else if (!Utils.isPasswordLengthMinimumFourChars(password)) {
            showValidationErrorMessage(
                    dataManager.getResourcesManager().getString(R.string.notEnoughPasswordChars)
            )
        } else {
            nextButtonVisibility.set(View.VISIBLE)
            errorVisibility.set(View.INVISIBLE)
        }
    }

    private fun showValidationErrorMessage(errorMessage: String) {
        validationPasswordErrorText.set(errorMessage)
        errorVisibility.set(View.VISIBLE)
        nextButtonVisibility.set(View.GONE)
    }

    fun startMainActivity(view: View) {
        dataManager.storeAuthenticationHeader(emailAddress)
        getNavigator()!!.moveForward(Navigator.Options.START_MAIN_ACTIVITY, emailAddress)
        getNavigator()!!.finish()
    }

    fun sendEmailToVerification(email: String) {
        emailAddress = email
    }

    fun checkIfUserIsVerified() {
        if (isNavigatorAttached) {
            dataManager.storeAuthenticationHeader(emailAddress)
            getNavigator()!!.moveForward(Navigator.Options.START_MAIN_ACTIVITY, emailAddress)
        }
    }

    companion object {

        private val START_EMAIL_ACTIVITY = 0
        private val START_PASSWORD_ACTIVITY = 1
    }

}